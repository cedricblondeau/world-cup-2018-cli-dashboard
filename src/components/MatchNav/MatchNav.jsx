import React from 'react';
import PropTypes from 'prop-types';

import MatchNavItem from './MatchNavItem/MatchNavItem';

const matchNavItemWidth = 23;

const getMatchNavItemKey = match => `match-nav-item-${match.fifa_id}`;

const getPaginatedMatchNavItems = (matches, selectedMatchIndex) => {
  const width = process.stdout.columns;
  const pageSize = Math.floor(width / matchNavItemWidth);

  const page = Math.floor(selectedMatchIndex / pageSize);
  const selectedIndex = selectedMatchIndex % pageSize;

  const sliceStart = page * pageSize;
  const matchesToDisplay = matches.slice(sliceStart, sliceStart + pageSize + 1);

  return matchesToDisplay.map((match, i) => (
    <MatchNavItem
      isSelected={selectedIndex === i}
      key={getMatchNavItemKey(match)}
      left={i * matchNavItemWidth}
      width={matchNavItemWidth}
      match={match}
    />
  ));
};

const getMatchNavItems = (matches, selectedMatchIndex = null) => {
  if (selectedMatchIndex === null) {
    return matches.map((match, i) => (
      <MatchNavItem
        isSelected={false}
        key={getMatchNavItemKey(match)}
        left={i * matchNavItemWidth}
        width={matchNavItemWidth}
        match={match}
      />
    ));
  }
  return getPaginatedMatchNavItems(matches, selectedMatchIndex);
};

const MatchNav = props => (
  <element height={5}>
    <box top={0} height={2} style={{ bg: 'black' }} scrollable>
      {getMatchNavItems(
        props.currentOrLastMatches,
        props.selectedPastOrCompletedMatchIndex,
      )}
    </box>
    <box top={3} height={2} scrollable>
      {getMatchNavItems(props.nextMatches)}
    </box>
  </element>
);

MatchNav.propTypes = {
  currentOrLastMatches: PropTypes.array.isRequired,
  nextMatches: PropTypes.array.isRequired,
  selectedPastOrCompletedMatchIndex: PropTypes.number.isRequired,
};

export default MatchNav;
