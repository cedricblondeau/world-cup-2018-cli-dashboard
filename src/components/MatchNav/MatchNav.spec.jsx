import React from 'react';
import TestRenderer from 'react-test-renderer';

import matchFixture from '../../../fixtures/match.json';
import MatchNav from './MatchNav';

const getTestMatch = status => {
  const match = Object.assign({}, matchFixture);
  match.status = status;
  match.fifa_id = Math.random()
    .toString(36)
    .substr(2, 9);
  return match;
};

const getTestMatches = (status = 'completed', count = 10) =>
  new Array(count).fill().map(() => getTestMatch(status));

const getMatchNavRendered = (
  currentOrLastMatchesCount = 10,
  nextMatchesCount = 10,
  selectedPastOrCompletedMatchIndex = 0,
) =>
  TestRenderer.create(
    <MatchNav
      currentOrLastMatches={getTestMatches(
        'in progress',
        currentOrLastMatchesCount,
      )}
      nextMatches={getTestMatches('completed', nextMatchesCount)}
      selectedPastOrCompletedMatchIndex={selectedPastOrCompletedMatchIndex}
    />,
  );

describe('<MatchNav>', () => {
  it('renders properly', () => {
    expect(getMatchNavRendered().toJSON()).toMatchSnapshot();
  });
});
