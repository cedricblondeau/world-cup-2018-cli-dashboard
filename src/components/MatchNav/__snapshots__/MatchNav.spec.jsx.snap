// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MatchNav> renders properly 1`] = `
<element
  height={5}
>
  <box
    height={2}
    scrollable={true}
    top={0}
  >
    <element
      height={2}
      left={0}
      top={0}
      width={23}
    >
      <box
        align="center"
        content="LIVE full-time"
        style={
          Object {
            "bg": "red",
            "fg": "white",
          }
        }
        top={0}
        width="100%"
      />
      <box
        align="center"
        content="FRA - ARG [30m[47m1/8[49m[39m"
        style={
          Object {
            "bg": "red",
            "fg": "white",
          }
        }
        tags={true}
        top={1}
        width="100%"
      />
    </element>
    <element
      height={2}
      left={23}
      top={0}
      width={23}
    >
      <box
        align="center"
        content="LIVE full-time"
        style={
          Object {
            "bg": "black",
            "fg": "white",
          }
        }
        top={0}
        width="100%"
      />
      <box
        align="center"
        content="FRA - ARG [30m[47m1/8[49m[39m"
        style={
          Object {
            "bg": "black",
            "fg": "white",
          }
        }
        tags={true}
        top={1}
        width="100%"
      />
    </element>
    <element
      height={2}
      left={46}
      top={0}
      width={23}
    >
      <box
        align="center"
        content="LIVE full-time"
        style={
          Object {
            "bg": "black",
            "fg": "white",
          }
        }
        top={0}
        width="100%"
      />
      <box
        align="center"
        content="FRA - ARG [30m[47m1/8[49m[39m"
        style={
          Object {
            "bg": "black",
            "fg": "white",
          }
        }
        tags={true}
        top={1}
        width="100%"
      />
    </element>
    <element
      height={2}
      left={69}
      top={0}
      width={23}
    >
      <box
        align="center"
        content="LIVE full-time"
        style={
          Object {
            "bg": "black",
            "fg": "white",
          }
        }
        top={0}
        width="100%"
      />
      <box
        align="center"
        content="FRA - ARG [30m[47m1/8[49m[39m"
        style={
          Object {
            "bg": "black",
            "fg": "white",
          }
        }
        tags={true}
        top={1}
        width="100%"
      />
    </element>
    <element
      height={2}
      left={92}
      top={0}
      width={23}
    >
      <box
        align="center"
        content="LIVE full-time"
        style={
          Object {
            "bg": "black",
            "fg": "white",
          }
        }
        top={0}
        width="100%"
      />
      <box
        align="center"
        content="FRA - ARG [30m[47m1/8[49m[39m"
        style={
          Object {
            "bg": "black",
            "fg": "white",
          }
        }
        tags={true}
        top={1}
        width="100%"
      />
    </element>
    <element
      height={2}
      left={115}
      top={0}
      width={23}
    >
      <box
        align="center"
        content="LIVE full-time"
        style={
          Object {
            "bg": "black",
            "fg": "white",
          }
        }
        top={0}
        width="100%"
      />
      <box
        align="center"
        content="FRA - ARG [30m[47m1/8[49m[39m"
        style={
          Object {
            "bg": "black",
            "fg": "white",
          }
        }
        tags={true}
        top={1}
        width="100%"
      />
    </element>
    <element
      height={2}
      left={138}
      top={0}
      width={23}
    >
      <box
        align="center"
        content="LIVE full-time"
        style={
          Object {
            "bg": "black",
            "fg": "white",
          }
        }
        top={0}
        width="100%"
      />
      <box
        align="center"
        content="FRA - ARG [30m[47m1/8[49m[39m"
        style={
          Object {
            "bg": "black",
            "fg": "white",
          }
        }
        tags={true}
        top={1}
        width="100%"
      />
    </element>
    <element
      height={2}
      left={161}
      top={0}
      width={23}
    >
      <box
        align="center"
        content="LIVE full-time"
        style={
          Object {
            "bg": "black",
            "fg": "white",
          }
        }
        top={0}
        width="100%"
      />
      <box
        align="center"
        content="FRA - ARG [30m[47m1/8[49m[39m"
        style={
          Object {
            "bg": "black",
            "fg": "white",
          }
        }
        tags={true}
        top={1}
        width="100%"
      />
    </element>
    <element
      height={2}
      left={184}
      top={0}
      width={23}
    >
      <box
        align="center"
        content="LIVE full-time"
        style={
          Object {
            "bg": "black",
            "fg": "white",
          }
        }
        top={0}
        width="100%"
      />
      <box
        align="center"
        content="FRA - ARG [30m[47m1/8[49m[39m"
        style={
          Object {
            "bg": "black",
            "fg": "white",
          }
        }
        tags={true}
        top={1}
        width="100%"
      />
    </element>
    <element
      height={2}
      left={207}
      top={0}
      width={23}
    >
      <box
        align="center"
        content="LIVE full-time"
        style={
          Object {
            "bg": "black",
            "fg": "white",
          }
        }
        top={0}
        width="100%"
      />
      <box
        align="center"
        content="FRA - ARG [30m[47m1/8[49m[39m"
        style={
          Object {
            "bg": "black",
            "fg": "white",
          }
        }
        tags={true}
        top={1}
        width="100%"
      />
    </element>
  </box>
  <box
    height={2}
    scrollable={true}
    top={3}
  >
    <element
      height={2}
      left={0}
      top={0}
      width={23}
    >
      <box
        align="center"
        content="06/30/2018 10:00 AM"
        style={
          Object {
            "bg": "black",
            "fg": "white",
          }
        }
        top={0}
        width="100%"
      />
      <box
        align="center"
        content="{bold}FRA{/bold} 4-3 ARG [30m[47m1/8[49m[39m"
        style={
          Object {
            "bg": "black",
            "fg": "white",
          }
        }
        tags={true}
        top={1}
        width="100%"
      />
    </element>
    <element
      height={2}
      left={23}
      top={0}
      width={23}
    >
      <box
        align="center"
        content="06/30/2018 10:00 AM"
        style={
          Object {
            "bg": "black",
            "fg": "white",
          }
        }
        top={0}
        width="100%"
      />
      <box
        align="center"
        content="{bold}FRA{/bold} 4-3 ARG [30m[47m1/8[49m[39m"
        style={
          Object {
            "bg": "black",
            "fg": "white",
          }
        }
        tags={true}
        top={1}
        width="100%"
      />
    </element>
    <element
      height={2}
      left={46}
      top={0}
      width={23}
    >
      <box
        align="center"
        content="06/30/2018 10:00 AM"
        style={
          Object {
            "bg": "black",
            "fg": "white",
          }
        }
        top={0}
        width="100%"
      />
      <box
        align="center"
        content="{bold}FRA{/bold} 4-3 ARG [30m[47m1/8[49m[39m"
        style={
          Object {
            "bg": "black",
            "fg": "white",
          }
        }
        tags={true}
        top={1}
        width="100%"
      />
    </element>
    <element
      height={2}
      left={69}
      top={0}
      width={23}
    >
      <box
        align="center"
        content="06/30/2018 10:00 AM"
        style={
          Object {
            "bg": "black",
            "fg": "white",
          }
        }
        top={0}
        width="100%"
      />
      <box
        align="center"
        content="{bold}FRA{/bold} 4-3 ARG [30m[47m1/8[49m[39m"
        style={
          Object {
            "bg": "black",
            "fg": "white",
          }
        }
        tags={true}
        top={1}
        width="100%"
      />
    </element>
    <element
      height={2}
      left={92}
      top={0}
      width={23}
    >
      <box
        align="center"
        content="06/30/2018 10:00 AM"
        style={
          Object {
            "bg": "black",
            "fg": "white",
          }
        }
        top={0}
        width="100%"
      />
      <box
        align="center"
        content="{bold}FRA{/bold} 4-3 ARG [30m[47m1/8[49m[39m"
        style={
          Object {
            "bg": "black",
            "fg": "white",
          }
        }
        tags={true}
        top={1}
        width="100%"
      />
    </element>
    <element
      height={2}
      left={115}
      top={0}
      width={23}
    >
      <box
        align="center"
        content="06/30/2018 10:00 AM"
        style={
          Object {
            "bg": "black",
            "fg": "white",
          }
        }
        top={0}
        width="100%"
      />
      <box
        align="center"
        content="{bold}FRA{/bold} 4-3 ARG [30m[47m1/8[49m[39m"
        style={
          Object {
            "bg": "black",
            "fg": "white",
          }
        }
        tags={true}
        top={1}
        width="100%"
      />
    </element>
    <element
      height={2}
      left={138}
      top={0}
      width={23}
    >
      <box
        align="center"
        content="06/30/2018 10:00 AM"
        style={
          Object {
            "bg": "black",
            "fg": "white",
          }
        }
        top={0}
        width="100%"
      />
      <box
        align="center"
        content="{bold}FRA{/bold} 4-3 ARG [30m[47m1/8[49m[39m"
        style={
          Object {
            "bg": "black",
            "fg": "white",
          }
        }
        tags={true}
        top={1}
        width="100%"
      />
    </element>
    <element
      height={2}
      left={161}
      top={0}
      width={23}
    >
      <box
        align="center"
        content="06/30/2018 10:00 AM"
        style={
          Object {
            "bg": "black",
            "fg": "white",
          }
        }
        top={0}
        width="100%"
      />
      <box
        align="center"
        content="{bold}FRA{/bold} 4-3 ARG [30m[47m1/8[49m[39m"
        style={
          Object {
            "bg": "black",
            "fg": "white",
          }
        }
        tags={true}
        top={1}
        width="100%"
      />
    </element>
    <element
      height={2}
      left={184}
      top={0}
      width={23}
    >
      <box
        align="center"
        content="06/30/2018 10:00 AM"
        style={
          Object {
            "bg": "black",
            "fg": "white",
          }
        }
        top={0}
        width="100%"
      />
      <box
        align="center"
        content="{bold}FRA{/bold} 4-3 ARG [30m[47m1/8[49m[39m"
        style={
          Object {
            "bg": "black",
            "fg": "white",
          }
        }
        tags={true}
        top={1}
        width="100%"
      />
    </element>
    <element
      height={2}
      left={207}
      top={0}
      width={23}
    >
      <box
        align="center"
        content="06/30/2018 10:00 AM"
        style={
          Object {
            "bg": "black",
            "fg": "white",
          }
        }
        top={0}
        width="100%"
      />
      <box
        align="center"
        content="{bold}FRA{/bold} 4-3 ARG [30m[47m1/8[49m[39m"
        style={
          Object {
            "bg": "black",
            "fg": "white",
          }
        }
        tags={true}
        top={1}
        width="100%"
      />
    </element>
  </box>
</element>
`;
